hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/protocol/datatransfer/sasl/SaslDataTransferServer.java

```diff
@@ -363,13 +366,31 @@ private IOStreamPair doSaslHandshake(Peer peer, OutputStream underlyingOut,
       InputStream underlyingIn, Map<String, String> saslProps,
       CallbackHandler callbackHandler) throws IOException {
 
-    DataInputStream in = new DataInputStream(underlyingIn);
+    DataInputStream in;
+    if (dnConf.getUnsafeDataTransferPlaintextFallback()) {
+      BufferedInputStream bufferedIn = new BufferedInputStream(underlyingIn);
+      in = new DataInputStream(bufferedIn);
+    } else {
+      in = new DataInputStream(underlyingIn);
+    }
     DataOutputStream out = new DataOutputStream(underlyingOut);
 
+    if (dnConf.getUnsafeDataTransferPlaintextFallback()) {
+      in.mark(4);
```

Comment:
For my own edification, what does this line do here?

Reply:
It tells the BufferedInputStream to remember at least 4 bytes, so that the later call to reset() can jump back that far